python/pyarrow/public-api.pxi

```diff
@@ -93,8 +98,8 @@ cdef api object pyarrow_wrap_data_type(
         out = Decimal128Type.__new__(Decimal128Type)
     elif type.get().id() == _Type_EXTENSION:
         ext_type = <const CExtensionType*> type.get()
-        if ext_type.extension_name() == PyExtensionName():
-            cpy_ext_type = <const CPyExtensionType*> ext_type
+        cpy_ext_type = dynamic_cast[_CPyExtensionTypePtr](ext_type)
+        if cpy_ext_type != nullptr:
```

Comment:
I didn't realize you could use `nullptr` in Cython is that new?

Reply:
> I didn't realize you could use nullptr in Cython is that new?

It's imported in lib.pxd with `from libcpp cimport nullptr`, and used in several places (but no idea if this is a new cython feature or not)